src/vm/gcenv.ee.cpp

```diff
@@ -830,53 +830,84 @@ void GCToEEInterface::StompWriteBarrier(WriteBarrierParameters* args)
         }
 #endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
 
-        stompWBCompleteActions |= ::StompWriteBarrierResize(args->is_runtime_suspended, args->requires_upper_bounds_check);
-
-        // We need to make sure that other threads executing checked write barriers
-        // will see the g_card_table update before g_lowest/highest_address updates.
-        // Otherwise, the checked write barrier may AV accessing the old card table
-        // with address that it does not cover.
-        //
-        // Even x86's total store ordering is insufficient here because threads reading
-        // g_card_table do so via the instruction cache, whereas g_lowest/highest_address
-        // are read via the data cache.
-        //
-        // The g_card_table update is covered by section 8.1.3 of the Intel Software
-        // Development Manual, Volume 3A (System Programming Guide, Part 1), about
-        // "cross-modifying code": We need all _executing_ threads to invalidate
-        // their instruction cache, which FlushProcessWriteBuffers achieves by sending
-        // an IPI (inter-process interrupt).
+        stompWBCompleteActions |= ::StompWriteBarrierResize(is_runtime_suspended, args->requires_upper_bounds_check);
+        is_runtime_suspended = (stompWBCompleteActions & SWB_EE_RESTART) || is_runtime_suspended;
 
         if (stompWBCompleteActions & SWB_ICACHE_FLUSH)
         {
-            // flushing icache on current processor (thread)
+            // flushing/invalidating the write barrier's body for the current process 
+            // NOTE: the underlying API may flush more than needed or nothing at all if Icache is coherent.
             ::FlushWriteBarrierInstructionCache();
-            // asking other processors (threads) to invalidate their icache
+        }
+
+        // NB: managed heap segments may surround unmanaged/stack segments. In such cases adding another managed heap segment
+        //     may put a stack/unmanaged write inside the new heap range. However the old card table would not cover it.
+        //     Therefore we must ensure that the write barriers see the new table before seeing new bounds.
+        //
+        //     On architectures with strong ordering, we only need to prevent compiler reordering.
+        //     Otherwise we put a process-wide fence here (so that we could use an ordinary read in the barrier)
+
+#if defined(_ARM64_) || defined(_ARM_)      
+        if (!is_runtime_suspended)
+        {
+            // If runtime is not suspended, force all threads to see the changed table before seeing updated heap boundaries.
+            // See:Dev11 #346765
```

Comment:
I do not think that "Dev11 #346765" is useful. Nobody will be able to find the bug using this reference. The internal bug database that this references is actually called DevDiv2. Either drop this note, or make it something that will allow folks to actually find the bug.

Reply:
I will add a link.

---
In reply to: [295927902](https://github.com/dotnet/coreclr/pull/25130#discussion_r295927902) [](ancestors = 295927902)